modules/custom/openy_repeat/src/Controller/RepeatController.php

```diff
@@ -0,0 +1,214 @@
+                AND
+                (re.start <= :timestamp_end)
+                AND
+                (re.end >= :timestamp_start)
+              )";
+
+    $values = [];
+    if (!empty($category)) {
+      $sql .= "AND re.category IN ( :categories[] )";
+      $values[':categories[]'] = explode(',', $category);
+    }
+    if (!empty($location)) {
+      $sql .= "AND nd.title IN ( :locations[] )";
+      $values[':locations[]'] = explode(',', $location);
+    }
+
+    $sql .= " ORDER BY re.start";
+
+    $values[':year'] = $year;
+    $values[':month'] = $month;
+    $values[':day'] = $day;
+    $values[':week'] = $week;
+    $values[':weekday'] = $weekday;
+    $values[':timestamp_start'] = $timestamp_start;
+    $values[':timestamp_end'] = $timestamp_end;
+
+    $connection = \Drupal::database();
+    $query = $connection->query($sql, $values);
+    $result = $query->fetchAll();
+
+    $locations_info = $this->getLocationsInfo();
+    $classes_info = $this->getClassesInfo();
+    foreach ($result as $key => $item) {
+      $result[$key]->location_info = $locations_info[$item->location];
+      $result[$key]->class_info = $classes_info[$item->class];
+    }
+
+    return new JsonResponse($result);
+  }
+
+  /**
+   * Return Location from "Session" node type.
+   *
+   * @return array
+   */
+  public function getLocations() {
+    $sql = "SELECT DISTINCT nd.title as location
+            FROM {node} n
+            INNER JOIN node__field_session_location l ON n.nid = l.entity_id AND l.bundle = 'session'
+            INNER JOIN node_field_data nd ON l.field_session_location_target_id = nd.nid
+            WHERE n.type = 'session'";
+
+    $connection = \Drupal::database();
+    $query = $connection->query($sql);
+
+    return $query->fetchCol();
+  }
+
+  /**
+   * Get detailed info about Location (aka branch).
+   */
+  public function getLocationsInfo() {
+    $nids = \Drupal::entityQuery('node')
+      ->condition('type','branch')
+      ->execute();
+    $branches = Node::loadMultiple($nids);
+
+    $data = [];
+    foreach ($branches as $node) {
+      $days = $node->get('field_branch_hours')->getValue();
+      $address = $node->get('field_location_address')->getValue();
+      if (!empty($address[0])) {
+        $address = array_filter($address[0]);
+        $address = implode(', ', $address);
+      }
+
+      $data[$node->title->value] = [
+        'nid' => $node->nid->value,
+        'title' => $node->title->value,
+        'email' => $node->field_location_email->value,
+        'phone' => $node->field_location_phone->value,
+        'address' => $address,
+        'days' => !empty($days[0]) ? $this->getFormattedHours($days[0]) : [],
+      ];
+    }
+
+    return $data;
+  }
+
+  /**
+   * Get detailed info about Class.
+   */
+  public function getClassesInfo() {
+    $nids = \Drupal::entityQuery('node')
+      ->condition('type','class')
+      ->execute();
+    $classes = Node::loadMultiple($nids);
```

Comment:
we can have tons of classes
loadmultiple is not an option, especially when you foreach array 
please splice array by 50 maximum